dxr/app.py

```diff
@@ -114,27 +122,36 @@ def _search_json(query, tree, query_text, is_case_sensitive, offset, limit, conf
             if is_case_sensitive:
                 params['case'] = 'true'
             return jsonify({'redirect': url_for('.browse', _anchor=line, **params)})
+
     try:
-        count_and_results = query.results(offset, limit)
-        # Convert to dicts for ease of manipulation in JS:
-        results = [{'icon': icon,
-                    'path': path,
-                    'lines': [{'line_number': nb, 'line': l} for nb, l in lines],
-                    'is_binary': is_binary}
-                   for icon, path, lines, is_binary in count_and_results['results']]
+        filters = query.instantiate_filters()
+        # Pull up all the non-negated text terms.
+        text_terms = [term for term in query.terms if term['name'] == 'text' and not term['not']]
```

Comment:
I should really have named it `_terms`, as I intended it to be private to `Query`. Can we pull the manipulation of `terms` into `Query`? In fact, isn't this just a duplication of the `Query.single_term()` logic `direct_results()` uses? Can `promoted_paths()` just call `single_term()` and, like `direct_result()`, return None if there isn't one?

Also, instead of requiring the caller to call instantiate_filters(), hang onto the result, and pass it into results() and promoted_paths()—thus making the filters a public data structure—why don't we just move filter instantiation to the Query constructor? We can keep the filters around in self.filters or wherever and reference them from the methods that need them.